### PR TITLE
fabtests: Add fi_rdm_multi_client test

### PR DESCRIPTION
**Pull request by shijin-aws**
This PR adds a new test `fi_rdm_multi_client`

This program tests the functionality of RDM endpoint in the case that a persistent server does ping-pong with multiple clients that come and leave in sequence. The client connects to a server, sends ping-pong, disconnects with the server by cleaning all fabric resources, and repeats. It will re-use the first client's address for the subsequent clients if the `fi_setname` API is implemented for the tested provider.

### CODE REVIEW

**Comment by shefty** [comment]
I started a review, but I don't think github is showing the commits in order.  Can you provide the order that the patches are applied, because I'm too lazy to pull down the PR and look myself?

**Review by bwbarrett — APPROVED** [review]

File: fabtests/prov/efa/src/rdm_cs.c
Comment (shefty) [inline fabtests/prov/efa/src/rdm_cs.c:178]: I don't see anything in this test that would make it provider specific.  Why isn't it just a new common test?

File: fabtests/common/shared.c
Comment (shefty) [inline fabtests/common/shared.c:771]: I would simplify this comment to describe the higher-level assumptions only.  Details like posting a receive, exchanging address, etc. can be determined directly from the code, and means if we change the code, we don't also need to remember to change the comment.  E.g.
```
Handles a persistent server communicating with multiple clients, one at a time, in sequence.
```

File: fabtests/common/shared.c
Comment (shefty) [inline fabtests/common/shared.c:792]: I would shrink this comment too to be concise.
```
Clients may be separate processes, so re-initialize any OOB setup.
```

File: fabtests/common/shared.c
Comment (shefty) [inline fabtests/common/shared.c:805]: Move this code into a new function, ft_reset_oob().  I could see how other tests might adopt a similar flow.

**Review by shefty — CHANGES_REQUESTED** [review]
Minor comments on the code changes.  The main request is that this doesn't appear to be a provider specific test, but one with general purpose.  Why should its use be restricted to a single provider?

I don't have a problem having provider specific tests as a rule.  It's just that the submitted test looks general to me.

**Comment by shefty** [comment]
Depending on what you're trying to stress, you could consider having the clients re-use the same address and not have the server update its AV.  In theory that should work.  In practice, we've discovered that it usually does not.

**Comment by shijin-aws** [comment]
That was actually the original motivation of this test. However, having the clients re-use the same address likely involves `fi_setname`, which has not been implemented in some provider (like efa) currently for rdm ep. It might have some value to check the return code of `fi_setname` to determine whether to force the address reuse. But I am not sure whether such branching is good for a common test. The current implementation is more like an extension to `functional/rdm.c` test, which actually only tests the server with 1 client.

**Comment by shefty** [comment]
I'm not asking for this change, but I think the test could support address re-use or not, depending on how it is run.  For example, if a source address is given for the client, that could signal to call fi_setname() and re-use the same source address.  (Or control this separately via some other option.)  Overall, I like this test and think it would be a nice addition which all providers could use.

**Comment by shijin-aws** [comment]
@shefty Thanks, I think that is a good idea. Currently when `-b` (OOB) is not specified for `runfabtests.sh`, source address is parsed into test executables as `-s <src_addr>`: https://github.com/ofiwg/libfabric/blob/master/fabtests/scripts/runfabtests.sh#L457-L471. I could re-use this option to make the test support address re-use or not.

**Comment by shijin-aws** [comment]
The updated PR still does not support client address re-use as I have not found a good way to pass string-format addresses to the test that is universal across multiple providers. I considered the `-s` option (`opts.src_addr`), but it seems to be the `node` in `fi_getinfo`.  Currently the only test that explicitly calls `fi_setname` is `functional/msg_sockets.c`, which passes the socket `ai.addr` of the `opts.src_addr` as `fi_setname`'s addr argument.

**Comment by shefty** [comment]
FWIW, I have a hacked up version of fi_rdm which tests multiple clients

https://github.com/shefty/libfabric/commit/0d7dfb4dee819ced23be77af42b1385442d846bd

I needed this to debug an issue.  The changes are close, but would still need some cleanup.

**Comment by shijin-aws** [comment]
@shefty Thanks for the example, this is helpful.

**Comment by shijin-aws** [comment]
The updated revision supports client's address reuse when `fi_setname` call is implemented for the tested provider. It will not reuse the address for subsequent clients if this API is not implemented.

File: fabtests/common/shared.c
Comment (shijin-aws) [reply fabtests/common/shared.c:767]: Fixed in newest revision.

File: fabtests/functional/rdm_multi_client.c
Comment (shefty) [inline fabtests/functional/rdm_multi_client.c:183]: Likewise, do we need the greeting?

File: fabtests/functional/rdm_multi_client.c
Comment (shijin-aws) [reply fabtests/functional/rdm_multi_client.c:93]: If I remove the greetings, I get a hang in ping-pong for client id 1. 
`ft_accept_next_client` calls `ft_post_rx`, which posts a recv, this recv gets completed in `ft_recv_meeting`, which calls `ft_get_rx_comp`. If I removes the `ft_post_rx` in `ft_accept_next_client`, the hang is resolved.

Comment (shefty) [reply fabtests/functional/rdm_multi_client.c:93]: I think we want to restructure the code to avoid the greeting, as it looks unnecessary.  That indicates there's subtleties in the underlying code which is making it more difficult to maintain.

The comment in ft_accept_next_client() indicates that the rx posting is to pick up the next client address.  If OOB address exchange is enabled, we shouldn't need to post that rx.  Maybe the ft_post_rx can be moved into an else clause of the OOB check.  We need to verify that fi_rdm doesn't break for nconn > 1.

Comment (shijin-aws) [reply fabtests/functional/rdm_multi_client.c:93]: I tried non-OOB path for tcp provider and the hang remains if I do not do greetings and keep the `ft_post_rx`... Is it expected?

**Comment by shijin-aws** [comment]
I looked at the AWS CI failures.. There are 3 types of failures.
1) single node shm provider hang for this test, backtrace
```
(gdb) bt
#0  0x00007f34eed0f0e3 in pthread_spin_lock ()
   from /lib64/libpthread.so.0
#1  0x00007f34efa1f236 in smr_verify_peer ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#2  0x00007f34efa1af4d in smr_generic_sendmsg ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#3  0x00007f34efa1b5c6 in smr_send ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#4  0x0000000000404694 in fi_send (context=0x6108c0 <tx_ctx>,
    dest_addr=0, desc=0x0, len=17, buf=0x1bd67c0, ep=0x1b96f80)
    at /fsx/libfabric/install/include/rdma/fi_endpoint.h:291
#5  ft_post_tx_buf (ep=ep@entry=0x1b96f80, fi_addr=fi_addr@entry=0,
    size=size@entry=17, data=data@entry=0,
    ctx=ctx@entry=0x6108c0 <tx_ctx>, op_buf=0x1bd67c0,
    op_mr_desc=0x0, op_tag=0) at common/shared.c:1908
#6  0x0000000000404825 in ft_post_tx (ep=ep@entry=0x1b96f80,
    fi_addr=fi_addr@entry=0, size=size@entry=17, data=data@entry=0,
    ctx=ctx@entry=0x6108c0 <tx_ctx>) at common/shared.c:1919
#7  0x0000000000407d25 in ft_tx (ep=0x1b96f80, fi_addr=0, size=17,
    ctx=0x6108c0 <tx_ctx>) at common/shared.c:1930
#8  0x0000000000409bac in ft_init_av_dst_addr (
    av_ptr=<optimized out>, ep_ptr=0x1b96f80,
    remote_addr=remote_addr@entry=0x610590 <remote_fi_addr>)
    at common/shared.c:1331
#9  0x0000000000409c40 in ft_init_av () at common/shared.c:1272
#10 0x00000000004017ec in run_client (client_id=1)
    at functional/rdm_multi_client.c:173
#11 main (argc=<optimized out>, argv=<optimized out>)
    at functional/rdm_multi_client.c:228
(gdb) quit
```
and I find the hang can be resolved by not doing address-reuse `fi_setname` for shm provider.

2) single node efa provider hang for this test,
3) multi node efa provider failed for this test.
I will look into the failures.
efa provider test does not have address-reuse due to `fi_setname` is not implemented. The failure

File: fabtests/functional/rdm_multi_client.c
Comment (shijin-aws) [reply fabtests/functional/rdm_multi_client.c]: Removed in the latest revision.

**Comment by shijin-aws** [comment]
The shm failure in earlier CI could be reproduced by fi_rdm with -C 2. So I created an issue #6794 . In this PR I exclude the test temporarily for shm provider. The efa failure in earlier CI is due to a bug in efa provider and we are fixing it in progress. So I also exclude this test for efa provider before the fix is merged.

**Comment by shijin-aws** [comment]
Travis CI shows the proposed test failed for sockets provider as well:
```
- name:   fi_rdm_multi_client -C 10 -I 5 -U -p "sockets"
  result: Fail
  time:   1
  server_cmd: fi_rdm_multi_client -C 10 -I 5 -U -p "sockets"  -s 127.0.0.1
  server_stdout: |
  client_cmd: fi_rdm_multi_client -C 10 -I 5 -U -p "sockets"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    run_client(): functional/rdm_multi_client.c:219, ret=22 (Unknown error -22)
```
Some other tests also failed 

```
- name:   fi_msg_pingpong -I 5 -v -p "sockets"
  result: Fail
  time:   3
  server_cmd: fi_msg_pingpong -I 5 -v -p "sockets"  -s 127.0.0.1
  server_stdout: |
    fi_listen(): common/shared.c:908, ret=-22 (Invalid argument)
  client_cmd: fi_msg_pingpong -I 5 -v -p "sockets"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    [error] fabtests:common/shared.c:2572: eq_readerr (Provider errno: -111) : Connection refused
- name:   fi_msg_bw -I 5 -p "sockets"
  result: Fail
  time:   2
  server_cmd: fi_msg_bw -I 5 -p "sockets"  -s 127.0.0.1
  server_stdout: |
    fi_listen(): common/shared.c:908, ret=-22 (Invalid argument)
  client_cmd: fi_msg_bw -I 5 -p "sockets"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    [error] fabtests:common/shared.c:2572: eq_readerr (Provider errno: -111) : Connection refused
```

@shefty could you help understand if the error is from the test or from the provider?

**Comment by shijin-aws** [comment]
Travis CI failed due  to `fi_rdm_multi_client` failed with udp provider when setting the first client's address to the second client via `fi_setname`.

```
 name:   fi_rdm_multi_client -C 10 -I 5 -p "udp"
  result: Fail
  time:   1
  server_cmd: fi_rdm_multi_client -C 10 -I 5 -p "udp"  -s 127.0.0.1
  server_stdout: |
  client_cmd: fi_rdm_multi_client -C 10 -I 5 -p "udp"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    Start ping-pong.
    Ping-pong succeeds.
    ft_setname(): functional/rdm_multi_client.c:149, ret=22 (Unknown error -22)
    run_client(): functional/rdm_multi_client.c:216, ret=22 (Unknown error -22)
    
```

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
@shefty Could you review the PR and leave feedback on the udp's `fi_setname` error?

**Comment by shefty** [comment]
The udp fi_setname() call goes direct to the standard socket bind() call:

```
udpx_setname(fid_t fid, void *addr, size_t addrlen)
...
	ret = bind(ep->sock, addr, (socklen_t)addrlen);
```
The return code in case of error will be the errno from bind().  Likewise, the fi_getname() is a direct call to getsockname()

**Comment by shefty** [comment]
The udp failure is "Address already in use".  Was the previous endpoint destroyed?  Maybe we need a socket option to reuse the address, though I didn't think that was needed for udp...

**Comment by shijin-aws** [comment]
@shefty Thanks for your comment!  The current test code calls `ft_free_res`, which will call `FI_CLOSE_FID(ep)`,  after each client finishes. I will check if the endpoint is destroyed correctly for udp.

**Comment by shijin-aws** [comment]
@shefty Hi, I debugged the udp error a bit,  but I think the root cause of the error is not the address in use? In travis CI the first failure is the invalid argument error returned by `fi_setname`.

```
name:   fi_rdm_multi_client -C 10 -I 5 -p "udp"
  result: Fail
  time:   1
  server_cmd: fi_rdm_multi_client -C 10 -I 5 -p "udp"  -s 127.0.0.1
  server_stdout: |
  client_cmd: fi_rdm_multi_client -C 10 -I 5 -p "udp"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    ft_setname(): functional/rdm_multi_client.c:149, ret=22 (Unknown error -22)
    run_client(): functional/rdm_multi_client.c:216, ret=22 (Unknown error -22)
    Start ping-pong.
    Ping-pong succeeds.
```

The `Address in use` error is from the subsequent `fi_rdm_multi_client -U` test, 
```
 name:   fi_rdm_multi_client -C 10 -I 5 -U -p "udp"
  result: Fail
  time:   1
  server_cmd: fi_rdm_multi_client -C 10 -I 5 -U -p "udp"  -s 127.0.0.1
  server_stdout: |
    fi_endpoint(): common/shared.c:681, ret=-98 (Address already in use)
    ft_init_fabric(): functional/rdm_multi_client.c:86, ret=98 (Unknown error -98)
    run_server(): functional/rdm_multi_client.c:225, ret=98 (Unknown error -98)
  client_cmd: fi_rdm_multi_client -C 10 -I 5 -U -p "udp"  -s 127.0.0.1 127.0.0.1
  client_stdout: |
    ft_setname(): functional/rdm_multi_client.c:149, ret=22 (Unknown error -22)
    run_client(): functional/rdm_multi_client.c:216, ret=22 (Unknown error -22)
    Start ping-pong.
    Ping-pong succeeds.
```
which should be due to the earlier `fi_rdm_multi_client` test failed but the server process has not been killed cleanly yet.

So the error is that when the `fi_setname` uses the address and sizes from `fi_getname` (for 1st client) to set name for 2nd client, the `fi_setname` returns -22. 

> Maybe we need a socket option to reuse the address, though I didn't think that was needed for udp...

Are you saying for udp we do not need to reuse the address for subsequent client?

**Comment by shefty** [comment]
As a start, the test should be fixed to handle the error codes correctly, and not print 'success' after seeing a failure.

Are you able to reproduce the error outside of the CI?  Based on code inspection, I don't see why fi_getname() passed into fi_setname() for the udp provider would fail.

Wait... this is using RDM eps, so RxD is getting used.  There may be a problem in that path.

**Comment by shefty** [comment]
RxD getname/setname calls pass directly through to the dgram provider.  I don't see anything odd in that path either.

**Comment by shijin-aws** [comment]
> As a start, the test should be fixed to handle the error codes correctly, and not print 'success' after seeing a failure.

The `success` message is for the first client. The failure happens for the second client. The output sequence is messed up for some reason. To clarify it, I added the client id in the printing in the newest revision.

> Are you able to reproduce the error outside of the CI? Based on code inspection, I don't see why fi_getname() passed into fi_setname() for the udp provider would fail. 

I could reproduce the error outside of the CI. 

> RxD getname/setname calls pass directly through to the dgram provider. I don't see anything odd in that path either.

May I know Is there an existing test in fabtests running this path? The only test I can see that uses `fi_setname` is `msg_sockets.c`

**Comment by shijin-aws** [comment]
Also, may I know what is the error in the CI "continuous-integration/jenkins/pr-head" ?

**Comment by shefty** [comment]
Intel CI failure is:
```
ame:   fi_rdm_multi_client -C 10 -I 5 -p "psm2"
17:27:01    result: Fail
17:27:01    time:   301
17:27:01    server_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6752/15/dbg/bin/fi_rdm_multi_client -C 10 -I 5 -p "psm2"  -s n5
17:27:01    server_stdout: |
17:27:01      Killed by signal 15.
17:27:01    client_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6752/15/dbg/bin/fi_rdm_multi_client -C 10 -I 5 -p "psm2"  -s n6 n5
17:27:01    client_stdout: |
17:27:01      n6.71639Received eager message(s) ptype=0x1 opcode=0xc9 from an unknown process (err=49)
17:27:01      n6.71639Received eager message(s) ptype=0x1 opcode=0xc9 from an unknown process (err=49)
<last msg repeated a bunch of times>
```
May just need to disable the test on psm2.

**Comment by shijin-aws** [comment]
For the `fi_setname` error with udp, I wrote a trivial test code outside fabtests to reproduce. See the attached `test_setname.txt`

I tried both `RDM` and `DGRAM` EP types, but get the same erro (-22) in `fi_setname` when using the same `name` and `namelen` obtained from `fi_getname`.
[test_setname.txt](https://github.com/ofiwg/libfabric/files/6606121/test_setname.txt)

**Comment by shefty** [comment]
The udp provider calls fi_setname() (bind) from fi_endpoint() creation, if the fi_info structure contains a src_addr.  I believe that's the reason fi_setname() is failing later, as an address has already been assigned.

udp could be modified to allow fi_setname() to change the address that was assigned at creation, but the current behavior is also aligned with the man pages.  So to be generic, I think in this case it would be better to update the test to handle this.

There may be a couple options here.  Either clear the src_addr when creating endpoints 2..n and still call fi_setname(), or set the src_addr for endpoints 2..n to the output of fi_getname().

**Comment by shijin-aws** [comment]
@shefty Thanks for the suggestions:

> There may be a couple options here. Either clear the src_addr when creating endpoints 2..n and still call fi_setname(), or set the src_addr for endpoints 2..n to the output of fi_getname().

I tried option 1) by setting `fi->src_addr = NULL; fi->src_addrlen = 0` before calling `fi_endpoint()`, but the NULL ptr of `fi->src_addr` is not passed to `udpx_endpoint()` because `rxd_endpoint()` is called first. `rxd_endpoint` initializes `dg_info` in `ofi_get_core_info` , which has a non-NULL src_addr for `dg_info`, then it `memcpy(dg_info->src_addr, info->src_addr, info->src_addrlen);` Since I have set `info->src_addrlen=0` this memcpy does not change `dg_info` so its src_addr is still non-NULL. Then `rxd_endpoint` calls `fi_endpoint` with dg_info, which passes a non-NULL src_addr to the `udpx_endpoint` call, and I hit the same error again.

For option 2), I applied this patch to my current PR

```
+       if (client_id > 0 && address_reuse) {
+               memcpy(fi->src_addr, name, size);
+               fi->src_addrlen = size;
+       }
+
        ret = ft_alloc_active_res(fi);
        if (ret) {
                FT_PRINTERR("ft_alloc_active_res", -ret);
                return ret;
        }

-       if (client_id > 0 && address_reuse) {
-               ret = fi_setname(&ep->fid, name, size);
-               if (ret) {
-                       FT_PRINTERR("fi_setname", -ret);
-                       return ret;
-               }
-       }
-
```
It works for tcp, sockets, but hit a hang with udp provider when reusing the address. 

The stack trace:
```
Client side:
#0  0x00007f186831b457 in recvmsg () from /lib64/libpthread.so.0
#1  0x00007f1868fcac77 in udpx_ep_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#2  0x00007f1868fa2185 in ofi_cq_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#3  0x00007f1868fa1686 in ofi_cq_readfrom ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#4  0x00007f186901d54f in rxd_ep_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#5  0x00007f1868fa2185 in ofi_cq_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#6  0x00007f1868fa1686 in ofi_cq_readfrom ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#7  0x0000000000405896 in fi_cq_read (count=1, buf=0x7ffea628b5e0,
    cq=0xbceff0) at /fsx/libfabric/install/include/rdma/fi_eq.h:394
#8  ft_spin_for_comp (timeout=-1, total=1, cur=0x610920 <rx_cq_cntr>,
    cq=0xbceff0) at common/shared.c:2232
#9  ft_get_cq_comp (cq=0xbceff0, cur=0x610920 <rx_cq_cntr>, total=1,
    timeout=-1) at common/shared.c:2324
#10 0x0000000000408309 in ft_rx (ep=0xc49c10, size=1)
    at common/shared.c:2173
#11 0x0000000000409b24 in ft_init_av_dst_addr (av_ptr=<optimized out>,
    ep_ptr=0xc49c10,
    remote_addr=remote_addr@entry=0x610590 <remote_fi_addr>)
    at common/shared.c:1335
#12 0x0000000000409ba0 in ft_init_av () at common/shared.c:1272
#13 0x000000000040178d in run_client (address_reuse=true, client_id=1)
    at functional/rdm_multi_client.c:159
#14 main (argc=<optimized out>, argv=<optimized out>)
    at functional/rdm_multi_client.c:228
(gdb) q


Server side

#0  0x00007f9040831457 in recvmsg () from /lib64/libpthread.so.0
#1  0x00007f90414e0c77 in udpx_ep_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#2  0x00007f90414b8185 in ofi_cq_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#3  0x00007f90414b7686 in ofi_cq_readfrom ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#4  0x00007f904153354f in rxd_ep_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#5  0x00007f90414b8185 in ofi_cq_progress ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#6  0x00007f90414b7686 in ofi_cq_readfrom ()
   from /fsx/libfabric/install/lib/libfabric.so.1
#7  0x0000000000405896 in fi_cq_read (count=1, buf=0x7ffc26e3ecd0,
    cq=0xb0eaa0) at /fsx/libfabric/install/include/rdma/fi_eq.h:394
#8  ft_spin_for_comp (timeout=-1, total=7, cur=0x610920 <rx_cq_cntr>,
    cq=0xb0eaa0) at common/shared.c:2232
#9  ft_get_cq_comp (cq=0xb0eaa0, cur=0x610920 <rx_cq_cntr>, total=7,
    timeout=-1) at common/shared.c:2324
#10 0x00000000004099ca in ft_init_av_dst_addr (av_ptr=0xb36570,
    ep_ptr=0xb3d830,
    remote_addr=remote_addr@entry=0x610590 <remote_fi_addr>)
    at common/shared.c:1339
#11 0x0000000000409ba0 in ft_init_av () at common/shared.c:1272
#12 0x0000000000401997 in run_server ()
```

**Comment by shijin-aws** [comment]
@shefty May I know what's the error in `continuous-integration/jenkins/pr-head` and any more changes required for this PR?

**Comment by shefty** [comment]
This is the error.  It may be thatyou just need to rebase against the latest main.  I thought there was a recent fix in this area.  I marked the PR as do not merge simply to delay until after v1.13 is released.  I'll re-review the latest changes.

```
17:41:56  - name:   fi_av_xfer -e rdm -p "udp;ofi_rxd"
17:41:56    result: Fail
17:41:56    time:   4
17:41:56    server_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6752/19/dbg/bin/fi_av_xfer -e rdm -p "udp;ofi_rxd"  -s n15
17:41:56    server_stdout: |
17:41:56      AV address removal: PASS
17:41:56      AV re-insertion address: fi_av_xfer: ./include/ofi_mem.h:443: ofi_bufpool_get_ibuf: Assertion `ofi_buf_region(buf)->use_cnt' failed.
17:41:56    client_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6752/19/dbg/bin/fi_av_xfer -e rdm -p "udp;ofi_rxd"  -s n16 n15
17:41:56    client_stdout: |
```

**Comment by shijin-aws** [comment]
@shefty thanks for the information. I will rebase the PR on top of main.

**Review by shefty — APPROVED** [review]
PR looks good to me.  thanks!

**Comment by shefty** [comment]
I had some old review comments queued up, which are no longer relevant, but ended up being added with my review.  You can ignore those.  The PR looks good as-is.

**Comment by shijin-aws** [comment]
@shefty Thanks, the updated the PR passes the Intel CI.

File: fabtests/prov/efa/src/rdm_cs.c
Comment (shijin-aws) [reply fabtests/prov/efa/src/rdm_cs.c]: Thanks, we finally made this test as a general test, so we do not need to set `prov_name` explicitly.

Comment (rajachan) [reply fabtests/prov/efa/src/rdm_cs.c]: 👍

**Review by rajachan — APPROVED** [review]

**Pull request merged**
